CONTRIBUTING.md

```diff
@@ -2,11 +2,66 @@
-- Once working, submit a PR to the GitHub actions repo.
+The projects mentioned in [README.md](README.md) are considered critical users of this repo.
+When introducing a change in this repo, care should be taken to check that those changes will not break the CI process of the critical users.
+
+To submit a change please follow the guidelines described below.
+
+## Basics
+### Pointing to a specific action
+The syntax for using a public action in a subdirectory is as follows ([reference](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory)):
+```
+uses: {owner}/{repo}/{path}@{ref}
+```
+
+For example:
+```
+uses: mfem/github-actions/build-hypre@v2.4
+```
+`{ref}` can be *any* of the following:
+- branch
+- tag
+- commit hash
+
+## Guidelines
+To submit a change please follow these guidelines.
+Because this involves making changes in multiple repos, each step will be marked with the repo it is referencing:
+ - `actions` refers to making a change in this repo: `mfem/github-actions`
+ - `project` refers to making a change in the repo with a workflow invoking an action, e.g. one of the projects mentioned in [README.md](README.md)
+
+### Steps
+1. `actions` Create a branch from `master` with a descriptive name, e.g. `add-single-precision-ci`
+2. `actions` Make updates to actions and push changes
+3. `project` Create a new branch and in that branch, change `{ref}` in the workflows you'd like to test so that it points to your feature branch, e.g.
+> `mfem/.github/workflows/build-and-tests.yml`
+> ```diff
+> - uses: mfem/github-actions/build-mfem@v2.4
+> + uses: mfem/github-actions/build-mfem@add-single-precision-ci
+> ```
+4. `project` To test the updated workflow, create a pull request and tag it with `DO-NOT-MERGE`; this will trigger the CI to run with your new changes
+
+> **Note**
+>
+> Adding an option to use manually trigger can be useful for debugging. This can be done by adding `workflow_dispatch` to the list of triggers
+> ```
+> on:
+>   workflow_dispatch
+> ```
+> The option to manually trigger will not appear until the workflow has previously been triggered via something like a PR or push.
+
+5. `actions` Continue to develop and test your action; because the `{ref}` in your `project` workflow is pointing to your branch, re-running the `project` CI will always point to the latest commit.
+
+6. `actions` Once your changes are finalized, submit a PR to pull your branch into `master`
+7. `actions` *Tag* the merged commit with a new [version number](https://semver.org/) `vX.Y`
+8. `project` Update the `{ref}` in your actions to point to the new version number, e.g.
+> `mfem/.github/workflows/build-and-tests.yml`
+> ```diff
+> - uses: mfem/github-actions/build-mfem@add-single-precision-ci
+> + uses: mfem/github-actions/build-mfem@v2.5
+> ```
+9. `project` Submit a PR with your updated `{ref}`
```

Comment:
Overall great! 👍  Just here, the PR was created in one of the previous steps, so this should be rather about removing the "do-not-merge" label and finishing the review process (merging the PR).
Secondly, it would be good to maybe add an optional extra step for the case when the names of the tests, which requires to change the setting of the CI in the project repo, updating the expected tests, like it happened in mfem/mfem#4262.

Reply:
@justinlaughlin I agree with @najlkin’s first statement.

The second statement about changing tests name was less clear to me. Based on the mentioned PR, @najlkin I think you meant changes of the job name in the CI. In which case, there is indeed an interesting point to be made:
A note could emphasize that when adding a new parameter to the actions, the developer may also want to add a new dimension to the CI job matrix in the projects CI. Care should be taken not to increase the number of jobs out of reason, which is why we prefer listing the jobs rather than increasing the matrix dimension for every parameter.

Reply:
That is also a good point, but I am not sure if we are going a bit too far with that matrix dimension, as it is somewhat specific to MFEM 🤔 . What I meant was that there is GitHub settings of expected CI tests, so when you change the test (job) name, there is a mismatch and the CI shows it is unfinished indefinitely. When we merged the PR and updated the settings of GitHub CI, the mismatch happens in the opposite regard to all other PRs (triggering CI with the old test names), so all PRs have to merge in `master`.

Reply:
Thank you for the feedback :).

Good catch on step 9 - I updated the wording. 

I added in a "Tips" section to include useful information that is complementary to the main "Steps". I added your comment about the job matrix there @adrienbernede. 

@najlkin I am looking through my own repo (since I have permission to browse the "Settings" there) and am not sure what setting you are referring to about expected CI tests. Is this under `Settings/Code and Automation/Actions/Runners`? When you say "names of tests" change, is that the name of the job spawned by the project workflow (unrelated to the name of the action)? Do you mind adding your suggestion to the "Tips" section (or let me know how to word it and I'll add it)? I think it is a good suggestion but I don't fully understand and am not sure how to word it 😅

Reply:
I have not done it by myself, so I do not know where exactly to find it, but [Tzanio did](https://github.com/mfem/mfem/pull/4262#issuecomment-2118986510). It is not fully related to changes of actions, but jobs need to have unique names, as @adrienbernede pointed out, so it is connected. And it causes a lot of havoc, so I think it should be mentioned.

Reply:
Thank you @najlkin I understand now. I *think* that is only something that can be done by a [MFEM Editor](https://github.com/orgs/mfem/teams/editors); I added a new step 10 implementing your suggestion but with this in mind.